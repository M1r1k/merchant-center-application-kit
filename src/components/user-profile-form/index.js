export { default } from './user-profile-form';
