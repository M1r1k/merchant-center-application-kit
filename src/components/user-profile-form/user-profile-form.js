import React from 'react';
import PropTypes from 'prop-types';
import { injectIntl, intlShape } from 'react-intl';
import { Formik } from 'formik';
import Spacings from '@commercetools-local/ui-kit/materials/spacings';
import PageBottomSpacer from '@commercetools-local/core/components/page-bottom-spacer';
import WarningSaveToolbar from '@commercetools-local/core/components/warning-save-toolbar';
import UserProfileGeneralInfoSubform from '../user-profile-general-info-subform';
import UserProfilePersonalSettingsSubform from '../user-profile-personal-settings-subform';
import validate from './validations';

export class UserProfileForm extends React.Component {
  render() {
    return (
      <Formik
        initialValues={this.props.initialValues}
        onSubmit={this.props.onSubmit}
        enableReinitialize={true}
        validate={validate}
        render={form => (
          <form>
            <Spacings.Inset>
              <Spacings.Stack scale="m">
                <UserProfileGeneralInfoSubform form={form} />
                <UserProfilePersonalSettingsSubform form={form} />
              </Spacings.Stack>
              <PageBottomSpacer />
              <WarningSaveToolbar
                onSave={form.handleSubmit}
                onCancel={form.handleReset}
                shouldWarnOnLeave={form.dirty}
                isToolbarVisible={form.dirty}
                isToolbarDisabled={form.isSubmitting}
              />
            </Spacings.Inset>
          </form>
        )}
      />
    );
  }
}

UserProfileForm.displayName = 'UserProfileForm';
UserProfileForm.propTypes = {
  initialValues: PropTypes.shape({
    email: PropTypes.string.isRequired,
    firstName: PropTypes.string.isRequired,
    lastName: PropTypes.string.isRequired,
    language: PropTypes.string.isRequired,
    timeZone: PropTypes.string,
  }).isRequired,
  onSubmit: PropTypes.func.isRequired,
  intl: intlShape.isRequired,
};

export default injectIntl(UserProfileForm);
