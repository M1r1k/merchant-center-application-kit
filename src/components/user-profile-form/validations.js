export const validate = values => {
  const errors = {};
  if (values.firstName.trim().length === 0) errors.firstNameMissing = true;
  if (values.lastName.trim().length === 0) errors.lastNameMissing = true;
  return errors;
};

export default validate;
