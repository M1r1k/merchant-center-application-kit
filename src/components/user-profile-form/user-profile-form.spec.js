import React from 'react';
import { shallow } from 'enzyme';
import { intlMock } from '@commercetools-local/test-utils';
import WarningSaveToolbar from '@commercetools-local/core/components/warning-save-toolbar';
import UserProfileGeneralInfoPanel from '../user-profile-general-info-panel';
import { UserProfileForm, validate } from './user-profile-form';

const createTestProps = props => ({
  initialValues: {
    version: 2,
    email: 'john@snow.got',
    firstName: 'John',
    lastName: 'Snow',
    language: 'en',
    timeZone: 'Europe/Berlin',
  },
  onSubmit: jest.fn(),
  intl: intlMock,
  ...props,
});

const createFormikProps = props => ({
  values: {
    version: 2,
    email: 'john@snow.got',
    firstName: 'John',
    lastName: 'Snow',
    language: 'en',
    timeZone: 'Europe/Berlin',
  },
  errors: {},
  touched: {},
  handleBlur: jest.fn(),
  handleChange: jest.fn(),
  handleSubmit: jest.fn(),
  isSubmitting: false,
  setFieldValue: jest.fn(),
  setFieldTouched: jest.fn(),
  resetForm: jest.fn(),
  dirty: false,
  ...props,
});

describe('validate', () => {
  describe('when the first name is missing', () => {
    it('should mark that as an error', () => {
      expect(validate({ firstName: '', lastName: 'Foo' })).toEqual({
        firstNameMissing: true,
      });
    });
  });
  describe('when the last name is missing', () => {
    it('should mark that as an error', () => {
      expect(validate({ firstName: 'Foo', lastName: '' })).toEqual({
        lastNameMissing: true,
      });
    });
  });
});

describe('rendering', () => {
  let props;
  let wrapper;
  beforeEach(() => {
    props = createTestProps();
    wrapper = shallow(<UserProfileForm {...props} />);
  });
  it('should ensure layout structure', () => {
    expect(wrapper).toMatchSnapshot();
  });
  describe('rendering form', () => {
    let formikProps;
    let formikWrapper;
    beforeEach(() => {
      formikProps = createFormikProps();
      formikWrapper = shallow(
        wrapper.find('Formik').prop('render')(formikProps)
      );
    });
    it('should render <UserProfileGeneralInfoPanel>', () => {
      expect(formikWrapper).toRender(UserProfileGeneralInfoPanel);
    });
  });
  describe('<WarningSaveToolbar>', () => {
    describe('when form is pristine', () => {
      let formikProps;
      let formikWrapper;
      beforeEach(() => {
        formikProps = createFormikProps({ touched: {}, dirty: false });
        formikWrapper = shallow(
          wrapper.find('Formik').prop('render')(formikProps)
        );
      });
      it('should not warn on leave', () => {
        expect(formikWrapper.find(WarningSaveToolbar)).toHaveProp(
          'shouldWarnOnLeave',
          false
        );
      });
      it('should not show save toolbar', () => {
        expect(formikWrapper.find(WarningSaveToolbar)).toHaveProp(
          'isToolbarVisible',
          false
        );
      });
    });
    describe('when form is dirty', () => {
      let formikProps;
      let formikWrapper;
      beforeEach(() => {
        formikProps = createFormikProps({
          touched: { name: true },
          dirty: true,
        });
        formikWrapper = shallow(
          wrapper.find('Formik').prop('render')(formikProps)
        );
      });
      it('should warn on leave', () => {
        expect(formikWrapper.find(WarningSaveToolbar)).toHaveProp(
          'shouldWarnOnLeave',
          true
        );
      });
      it('should show save toolbar', () => {
        expect(formikWrapper.find(WarningSaveToolbar)).toHaveProp(
          'isToolbarVisible',
          true
        );
      });
    });
    describe('when form is submitting', () => {
      let formikProps;
      let formikWrapper;
      beforeEach(() => {
        formikProps = createFormikProps({
          touched: { firstName: true },
          dirty: true,
          isSubmitting: true,
        });
        formikWrapper = shallow(
          wrapper.find('Formik').prop('render')(formikProps)
        );
      });
      it('should disable SaveToolbar', () => {
        expect(formikWrapper.find(WarningSaveToolbar)).toHaveProp(
          'isToolbarDisabled',
          true
        );
      });
    });
    describe('when firstname is invalid', () => {
      let formikProps;
      let formikWrapper;
      beforeEach(() => {
        formikProps = createFormikProps({
          errors: { firstName: 'foo' },
          touched: { firstName: true },
          dirty: true,
        });
        formikWrapper = shallow(
          wrapper.find('Formik').prop('render')(formikProps)
        );
      });
      it('should show the error', () => {
        expect(formikWrapper.find(UserProfileGeneralInfoPanel)).toHaveProp(
          'errors',
          { firstName: 'foo' }
        );
      });
      it('should not disable SaveToolbar', () => {
        expect(formikWrapper.find(WarningSaveToolbar)).toHaveProp(
          'isToolbarDisabled',
          false
        );
      });
    });
  });
});
