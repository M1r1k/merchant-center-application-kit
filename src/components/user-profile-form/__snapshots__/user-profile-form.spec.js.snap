// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering should ensure layout structure 1`] = `
<Inset
  scale="m"
>
  <Stack
    alignItems="stretch"
    scale="m"
  >
    <InjectIntl(UserProfileGeneralInfoPanel)
      hasSubmitFailed={false}
    />
    <Connect(withTimeZones(InjectIntl(UserProfilePersonalSettingsPanel))) />
  </Stack>
  <PageBottomSpacer />
  <SaveToolbar
    isToolbarDisabled={false}
    isToolbarVisible={false}
    isVisible={false}
    onCancel={[MockFunction]}
    onSave={[MockFunction]}
    route={Object {}}
    shouldWarnOnLeave={false}
  />
</Inset>
`;
